pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+	}
+	return fmt.Sprintf("Union<%v>", strings.Join(elements, ", "))
+}
+
+func (*UnionType) isType() {}
+
+// ObjectType represents schematized maps from strings to particular types.
+type ObjectType struct {
+	// Token is the type's Pulumi type token.
+	Token string
+	// Comment is the description of the type, if any.
+	Comment string
+	// Properties is the list of the type's properties.
+	Properties []*Property
+	// Language specifies additional language-specific data about the object type.
+	Language map[string]json.RawMessage
+}
+
+func (t *ObjectType) String() string {
+	return t.Token
+}
+
+func (*ObjectType) isType() {}
+
+// TokenType represents an opaque type that is referred to only by its token. A TokenType may have an underlying type
+// that can be used in place of the token.
+type TokenType struct {
+	// Token is the type's Pulumi type token.
+	Token string
+	// Underlying type is the type's underlying type, if any.
+	UnderlyingType Type
+}
+
+func (t *TokenType) String() string {
+	return t.Token
+}
+
+func (*TokenType) isType() {}
+
+// DefaultValue describes a default value for a property.
+type DefaultValue struct {
+	// Value specifies a static default value, if any. This value must be representable in the Pulumi schema type
+	// system, and its type must be assignable to that of the property to which the default applies.
+	Value interface{}
+	// Environment specifies a set of environment variables to probe for a default value.
+	Environment []string
+	// Language specifies additional language-specific data about the default value.
+	Language map[string]json.RawMessage
+}
+
+// Property describes an object or resource property.
+type Property struct {
+	// Name is the name of the property.
+	Name string
+	// Comment is the description of the property, if any.
+	Comment string
+	// Type is the type of the property.
+	Type Type
+	// DefaultValue is the default value for the property, if any.
+	DefaultValue *DefaultValue
+	// IsRequired is true if the property must always be populated.
+	IsRequired bool
+	// Language specifies additional language-specific data about the property.
+	Language map[string]json.RawMessage
+}
+
+// Alias describes an alias for a Pulumi resource.
+type Alias struct {
+	// Name is the "name" portion of the alias, if any.
+	Name *string
+	// Project is the "project" portion of the alias, if any.
+	Project *string
+	// Type is the "type" portion of the alias, if any.
+	Type *string
+}
+
+// Resource describes a Pulumi resource.
+type Resource struct {
+	// Token is the resource's Pulumi type token.
+	Token string
+	// Comment is the description of the resource, if any.
+	Comment string
+	// IsProvider is true if the resource is a provider resource.
+	IsProvider bool
+	// InputProperties is the list of the resource's input properties.
+	InputProperties []*Property
+	// Properties is the list of the resource's output properties. This should be a superset of the input properties.
+	Properties []*Property
```

Comment:
1. should we potentially call this OutputProperties to be very explicit?
2. I didn't realizee the bit about the superset.  is that really always the case?